src/main/java/com/artipie/asto/OneTimePublisher.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 artipie.com
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.asto;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+
+/**
+ * A publish which can be consumed only once.
+ * @param <T> The type of publisher elements.
+ * @since 0.23
+ * @todo #204:30min Use one time publisher in Storage.
+ *  We need to wrap all publishers created in Storage implementations with this one
+ *  and to wrap all incoming publishers in tests.
+ */
+public final class OneTimePublisher<T> implements Publisher<T> {
+
+    /**
+     * The original publisher.
+     */
+    private final Publisher<T> original;
+
+    /**
+     * The amount of subscribers.
+     */
+    private final AtomicInteger subscribers;
+
+    /**
+     * Wrap a publish in a way it can be used only once.
+     * @param original The original publisher.
+     */
+    public OneTimePublisher(final Publisher<T> original) {
+        this.original = original;
+        this.subscribers = new AtomicInteger(0);
+    }
+
+    @Override
+    public void subscribe(final Subscriber<? super T> sub) {
+        final int subs = this.subscribers.incrementAndGet();
+        if (subs == 1) {
+            this.original.subscribe(sub);
+        } else {
+            final String msg =
+                "The subscriber could not be consumed more than once. Failed on #%d attempt";
+            sub.onError(new IllegalStateException(String.format(msg, subs)));
```

Comment:
@Sammers21 there are few ReactiveStreams [specification](https://github.com/reactive-streams/reactive-streams-jvm#specification) violations:
> 1.3 onSubscribe, onNext, onError and onComplete signaled to a Subscriber MUST be signaled serially.

> 1.8 If a Subscription is cancelled its Subscriber MUST eventually stop being signaled.

> 1.9 Publisher.subscribe MUST call onSubscribe on the provided Subscriber prior to any other signals to that Subscriber and MUST return normally, except when the provided Subscriber is null in which case it MUST throw a java.lang.NullPointerException to the caller, for all other situations the only legal way to signal failure (or reject the Subscriber) is by calling onError (after calling onSubscribe).

To fix it you have to call `sub.onSubscribe` before calling `onError` and provide some subscription implementation, then check if this subscription was not cancelled (because subscriber may cancel subscription in `onSubscribe`), and call `onError` if it was not cancelled. Also, you need to check if `sub` is null and throw `NullPointerException` in that case:
```java
Objects.requireNotNull(sub);
Subscription s = new SomeSubscription();
sub.onSubscribe(s);
if (!s.cancelled()) {
  sub.onError(err);
}
```

Reply:
I've made the wrapper tck compatible.